### PR TITLE
Remove deprecated versions from packages

### PR DESCRIPTION
**Pull request by adamjstewart**
For now, this only includes packages that I personally maintain until @tgamblin lets me do all packages. Let's see how much of CI breaks!

Notable removals:

* Anaconda 2
* Catalyst
* Ancient numpy/scipy
* Ancient PyTorch
* Ancient Bazel/TF

### CODE REVIEW

**Comment by haampie** [comment]
> Python 3.7

Let's not do this... I've explained it dozens of times.

**Review by haampie — DISMISSED** [review]
See above

File: var/spack/repos/builtin/packages/python/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/python/package.py]: Literally nothing is gained by removing this, no simplifications of the build instructions or whatever. This change solely complicates bootstrapping

**Comment by adamjstewart** [comment]
If you would like to retain Python 3.7 support, we're always looking for new maintainers for the Python recipe!

There may be no simplifications in the Python recipe, but there are still 3.5K lines of code removed as a result. Python 3.7 is EOL and may or may not contain known security vulnerabilities. I don't think Spack should install it when there are plenty of modern versions of Python that users can use instead.

**Comment by adamjstewart** [comment]
To go a step farther, I don't think Spack should support _running_ on Python 3.6 or 3.7. Users on ancient systems with ancient system Python can use an ancient version of Spack (circa 2022) or use that ancient version of Spack to bootstrap a modern version of Python and then upgrade Spack. But that's another discussion. Here I'm only limiting what users are able to _install_ with Spack.

**Comment by eli-schwartz** [comment]
> There may be no simplifications in the Python recipe, but there are still 3.5K lines of code removed as a result.

The PR as a whole is 3.5k lines of code but it doesn't appear like much of any of that has to do with python 3.7? Given the argument was specific to that one recipe, the defense of removing it should probably talk about how many lines of code removing "all packages other than that" vs. "removing that too" results in.

I could hear the argument that the python interpreter in particular is special and a nice-to-have for old (even EOL) versions. I myself have some interest in python 3.7 (not via spack, though) for e.g. bisection, regression testing, comparative analysis. The fact that it's EOL doesn't matter for those purposes.

...

My sole actual stake here is py-meson-python, and I am +9999 on removing a `.pre0` release that has an accompanying comment saying to remove it as soon as the final release occurs -- which happened a long time ago and the motivating factor in adding that release was scikit-image, which bumped its minimum requirement to the final release back in May and cannot use this `.pre0` version anyway.

There's no good excuse for end users to want to install prereleases after a final release exists.

**Comment by haampie** [comment]
- EOL according to upstream != EOL according to distros. Distros do (security) fixes regardless, people pay for that.
- `spack install x` will fail for people marking system Python external, since you've removing all compat rules instead of replacing them with `conflicts("python@:3.7")`. There's an overhead to dealing with those reports, which I think is larger than the "visual noise" of some depends_on and conflicts statements.
- The vast majority of the rules you're removing is of the type `depends_on("python@:3.7", when="@:x.y")`. These rules will never ever change anymore. There's no maintenance overhead in keeping them.
- Keeping Python 3.7 available makes it easier for us Spack developers to distribute binaries using the 3.7 abi. Without we have to special case it, it's a mess. 
- (Review of this PR itself is actually an example of maintenance overhead of removing Python 3.7)

**Comment by adamjstewart** [comment]
> EOL according to upstream != EOL according to distros. Distros do security fixes regardless, people pay for that.

I am not a distro, I'm a busy postdoc who doesn't have time to support ancient software. Again, happy to add new maintainers to the recipe if you're interested.

> There's an overhead to dealing with those reports, which I think is larger than the "visual noise" of some depends_on and conflicts statements.

I haven't yet received such a report, but I'm sure I'll be very busy dealing with it if it ever happens.

> The vast majority of the rules you're removing is of the type depends_on("python@:3.7", when="@:x.y").

I actually haven't removed those kinds of rules yet. That's another 1.5K+ lines of code if you want me to remove those too.

**Comment by adamjstewart** [comment]
> The PR as a whole is 3.5k lines of code but it doesn't appear like much of any of that has to do with python 3.7?

Almost everything is 3.7. It results in ancient versions of numpy/scipy/azureml/tf/pytorch being removed, which cascades through other packages. The only thing that isn't is bazel, which is no longer needed because we're removing ancient versions of TF because they require Python 3.7.

**Comment by haampie** [comment]
> I actually haven't removed those kinds of rules yet. That's another 1.5K+ lines of code if you want me to remove those too.

This PR removes a bunch of those lines still. Edit: ah, I see that they're sometimes dropped based on when condition no longer possible, but not always.

But take `paraview` as an example. What's gained with your changes?

> I'm a busy postdoc who doesn't have time to support ancient software

But in what way are you supporting ancient software? Those compat rules have been contributed already, you're not actively changing depends_on / conflicts for 3.7 support for new contributions? You also don't have to backport security fixes for Python 3.7 cause nobody is paying you for it ;p.

**Comment by eli-schwartz** [comment]
Sounds like as a first pass, just removing ancient versions of numpy/scipy/azureml/tf/pytorch might help quite a bit. In particular, I'd suggest looking at dropping "everything other than the most recent version which still supports python 3.7".

For any conceivable reason why someone would want to use python 3.7, I doubt they also have a compelling reason to use python modules which were old even when python 3.7 was state of the art. It's certainly not a requirement in order to use external system-python.

File: var/spack/repos/builtin/packages/python/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/python/package.py]: I kind of agree with the statement above. When preparing a new set of binaries for bootstrapping in https://github.com/spack/spack-bootstrap-mirrors, testing on a pristine Python 3.6 was more difficult than needed because the Spack under test couldn't install Python 3.6.

> To go a step farther, I don't think Spack should support running on Python 3.6 or 3.7

As you say this is another, related, discussion. For what is worth, if we decided that we don't support running on Python 3.7 all this would be fine. But I think removing this support and deleting the latest Python 3.7 version should go together.

**Comment by adamjstewart** [comment]
I undid changes to Python 3.7 because it is controversial. Let's discuss that in a separate PR after this is merged.

**Review by wdconinc — DISMISSED** [review]
Approve of the `gaudi` changes.

**Review by jmcarcell — APPROVED** [review]

**Comment by haampie** [comment]
> Let's discuss that in a separate PR after this is merged.

:confused: 

I'm not a fan of this PR at all, you didn't address my comments:

> But take paraview as an example. What's gained with your changes?

> But in what way are you supporting ancient software? Those compat rules have been contributed already, you're not actively changing depends_on / conflicts for 3.7 support for new contributions?

**Comment by adamjstewart** [comment]
> But take paraview as an example. What's gained with your changes?

@danlipsa @kwryankrattiger @vicentebolea If any of you would like to maintain Python 3.7 support in Spack or have any issues with the changes to the paraview package, let me know!

> But in what way are you supporting ancient software? Those compat rules have been contributed already, you're not actively changing depends_on / conflicts for 3.7 support for new contributions?

@manuelakuhn and I review hundreds of Python PRs every month. Every time we need to manually ensure that the dependencies listed are correct for every version of that package and every version of Python. Then come the several rounds of iteration in which we explain to contributors why they need to specify that this dependency used to be required, but is now only required for Python versions that are EOL but we're still supporting for some reason. The only way to avoid that is to draw a line somewhere and decide which versions of Python we're willing to support. This also allows me to deprecate and remove ancient versions of packages like numpy/scipy/TF/PyTorch at the same time, removing ~3.5K lines of code and packages no one cares about _every year_.

All changes in this PR are necessary just for CI to pass without Python 3.7. I haven't yet removed the thousands of lines of code like:
```python
depends_on("python@2.6:2.7,3.3:")
```
That's for another future PR 😄 

It's not entirely clear to me why these changes are controversial. But I'll add an FAQ just to clear some things up.

### FAQ

**Q: What if I want to keep Python 3.7 in Spack?**

A: Add yourself as a maintainer to the Python package.

**Q: What if I want to be able to install XYZ using Python 3.7 in Spack?**

A: Add yourself as a maintainer to the XYZ package and the Python package.

**Q: What if I want to be able to install all Python packages with Python 3.7?**

A: Add yourself as a maintainer to the PythonPackage base class and the Python package.

**Q: What if I don't want to maintain any of these packages but I still want someone else to?**

A: Remove me as a maintainer from these 2.4K packages.

**Comment by eli-schwartz** [comment]
Is it possible to make any package using python that doesn't explicitly specify it depends on python 3.7+, default to depending on 3.8+ (or whatever the minimum default python is, specified as a global constant)?

That might solve the issue of reviewing packages and asking people to manually handle 3.7, because no packages would support it unless the author goes out of their way to do so. But it also wouldn't impact whether really old python versions can be installed at all, or handle modules whose active developers really care about it for some reason.

You'd also be able to bump that once a year in a one-line PR. :D

**Comment by vicentebolea** [comment]
@adamjstewart no need to keep python3.7 dependency for Paraview. The changes looks good

**Review by vicentebolea — APPROVED** [review]
Paraview changes

**Comment by greenc-FNAL** [comment]
@adamjstewart Please note that there are several packages—I'm thinking specifically of `glib` and `gobject-introspection`, but a colleague has discovered at least one more—where the default build system has changed from Autotools to Meson and the former versions have been deprecated. Unfortunately this has led to a loss of functionality: neither of the two packages mentioned are able to deal with the `iconv`/`libintl`/`gettext`/`libc` issue.

I understand that these packages are not part of the current raft of removals, but you did mention that this was the thin end of the wedge.

**Comment by adamjstewart** [comment]
Duly noted. I would personally prefer to un-deprecate those if they are still needed than to remove them or keep them deprecated. These were deprecated before the multi-build system PR that made it much easier to manage those kinds of changes. But I'm not maintaining those, so it's up to @michaelkuhn or anyone else who wants to volunteer to be a maintainer.

**Comment by eli-schwartz** [comment]
> @adamjstewart Please note that there are several packages—I'm thinking specifically of `glib` and `gobject-introspection`, but a colleague has discovered at least one more—where the default build system has changed from Autotools to Meson and the former versions have been deprecated. Unfortunately this has led to a loss of functionality: neither of the two packages mentioned are able to deal with the `iconv`/`libintl`/`gettext`/`libc` issue.
> 
> I understand that these packages are not part of the current raft of removals, but you did mention that this was the thin end of the wedge.

I don't really have a stake in what versions of python spack supports. But I would very much love to hear more about your iconv/intl issues.

Speaking with my meson upstream developer hat on, it's much much much better to report issues rather than suffer in silence and pin to old versions of glib, including attendant issues like new software that you would like to use requiring new APIs. ***We want to help you by making meson better***.

We.have builtin iconv and intl handling. It is tested in CI and works in all of our test environments. Clearly, you have scenarios that we don't have a GitHub Actions runner for. (Or you are using old versions of meson.) What is your problem and how can we help you out here? 

Please throw me a bone. :) I don't know what "the `iconv`/`libintl`/`gettext`/`libc` issue" because the use of the word "the" along with no citations indicates you think it's a famous issue that no one needs to be told about -- but it's not famous to me.

**Comment by greenc-FNAL** [comment]
@eli-schwartz Many thanks for weighing in here.

At Fermilab my group (supporting experiments other than CMS) generally uses Spack stacks that have a reliance on the presence a certain set of external system packages, and on the SL7 (and more recently AL9) platform(s). This can lead to us seeing issues that don't come up in other people's CI testing.

[ Please forgive me if any of the below is already familiar to you, but it helps me to lay out everything so I can map and audit my own reasoning here. ]

In particular: Spack has a virtual `iconv` package whose requirements can either be satisfied by `libiconv` or `libc`. Separately but related, `libc` can also provide `libintl` functionality which is otherwise provided by specific configurations of `gettext`. A quick search for issues and PRs referring to `iconv`, `libiconv`, `libintl` or `gettext` will show that quite a few packages need help in their recipes to handle this situation.

When I was updating our local Spack fork to sync with upstream recently, I hit `libintl`-related build issues with `glib` and `gobject-introspection`. As I was trying to make progress on the main task (syncing our fork and continuing to be able to build our stacks), I did the minimum possible which was patch the recipes in our fork to remove the `deprecated` tag on the most recent non-meson builds and specify `buildsystem=autotools` in order to proceed.

There's also a minor issue with `jsoncpp`, since the meson build doesn't produce the CMake config file (not your problem, I know). This affects our ability to find `jsoncpp` from CMake builds of our own software which previously relied on `find_package()` doing the right thing due to having the CMake config file available.

I'd be happy to work with you on these problems in a more appropriate forum. Should I put issues in for the `glib` and `gobject-introspection` recipes, and maybe discuss `jsoncpp` on Slack? Or if you'd like to proceed another way, that's fine too.

**Comment by eli-schwartz** [comment]
Sure, feel free to @ me in a new issue(s) or discussion (looks like spack does have those enabled, and that works better for me since I'm not on Slack).


Quick advance pointers:

- double checking when some PRs were merged.. try to make sure you're testing against at least glib 2.73.1
- the jsoncpp thing could be solved using cmake's `find_package(PkgConfig)` and `pkg_check_modules(JSONCPP jsoncpp IMPORTED_TARGET GLOBAL)`

**Comment by tgamblin** [comment]
Ok folks I was asked to weigh in on this one in the absence of consensus bt/w @adamjstewart and @haampie.  So here's the plan, based on a discussion among the three of us:

1. We're going to keep Python 3.6+ in the `python` package, and @haampie is going to help maintain that.
2. Spack `builtin` will, for now, have *supported* Python versions following the [Status of Python Versions](https://devguide.python.org/versions/) page.  What that means for Spack is that the package ecosystem is supported for those versions (but `python` will be installable for Python 3.6+ because we need it to handle externals and bootstrap RHEL.
3. `clingo`, `gpg`, and other packages needed for spack itself to function properly still need to work with 3.6+, but nothing else.
4. That means that `spack style` and other commands that require us to bootstrap other python packages will only be supported for supported `python` versions, and `spack bootstrap` will fail if you try to bootstrap `--dev` dependencies for unsupported python versions.
5. @adamjstewart is going to revert changes that were made only to appease CI because older `python` versions were removed (i.e. no lower bounds when not needed, etc.), then we can merge the PR.

Finally, @haampie and us core spack developers are going to look into things like [PEP 658](https://peps.python.org/pep-0658/) (tracked by https://github.com/pypi/warehouse/issues/8254) and possibly indexing metadata via the PyPI [BigQuery metadata table](https://warehouse.pypa.io/api-reference/bigquery-datasets.html#project-metadata-table).  The hope is that in the long term we can get rid of pure python packages in Spack.  Specifically that would require us to, in the long term:
* Move python packages to their own repo, e.g. with a namespace like `py.whatever`
* Enable [Spack repositories](https://spack.readthedocs.io/en/latest/repositories.html) to do transparently fetch PyPI or other foreign package metadata behind the scenes, and handle all pure python packages that way.
* Add specific "traditional" spack packages for things that need native builds (e.g., TF, torch, ParaView, etc.)
* Add support for mirroring metadata so that we can still build python stuff over an air gap.

What this would do in the long term is to lean on what the actual Python ecosystem is going, and remove the maintenance burden of pure python packages (which are an awful lot of PRs).  Then folks like @adamjstewart can focus on, say, ML and geoscience packages instead of merging every tiny Python PR here.  If we get this working, you could see historical python packages come back in Spack.  We could also similarly adapt ecosystems like rust, go, npm, etc., as this foreign package ecosystem maintenance problem is not unique to Python.

This is long-term, but in the meantime we have the compromise above.  If folks *still* want older python versions in the meantime, I think the solution is to start a separate Spack repo for those, e.g. `py-legacy`. This is not unlike what other projects like Homebrew have done with taps, and I'm not opposed to making it simpler to register hosted package repositories in Spack (right now we annoyingly have to clone them separately and Spack doesn't manage it).  If you want that, let us know.

Thanks to all the maintainers who have commented here for your contributions to Spack (and other projects we use).... I hope we can get core features built fast enough to keep up. Packaging just keeps getting harder.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
